### ISSUE TITLE
Where are the 'to' functions?

### ISSUE THREAD

**Issue by gazzatav** [taocpp/json#103]
Following on from the discussion 'Beginners luck running out - help needed accessing values.
#102'

It seemed to be a simple typo, however this was introduced while redacting and organising the code for display on github.

With the missing '>' mine on g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0 still does not compile with error:
```
expt2.cpp: In function ‘int main(int, const char**)’:
expt2.cpp:31:26: error: ‘std::map<std::__cxx11::basic_string<char>, tao::json::basic_value<tao::json::traits>, std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, tao::json::basic_value<tao::json::traits> > > >::mapped_type {aka class tao::json::basic_value<tao::json::traits>}’ has no member named ‘to’
       meta.at("version").to<std::string>(version);
                          ^~
expt2.cpp:31:40: error: expected primary-expression before ‘>’ token
       meta.at("version").to<std::string>(version);
```

_Originally posted by @gazzatav in https://github.com/taocpp/json/discussions/102#discussioncomment-781775_

I have grepped both the cloned repo and my conan cache for " to<" and many other variants and all I can find (that is close) is 'wrap_to' in binding/internal/object.hpp.  So at the moment I agree with the compiler - there is no member function 'to'!  My understanding is that 'meta.at("version")' returns a tao::json::basic_value< traits etc'  I thought I should be able to find several specialisations of 'to' in traits.  What am I missing?

**Comment by d-frey** [comment]
The `to` overloads can be found here: https://github.com/taocpp/json/blob/0dc7bd6984d26cb7c0c0d37d6714398dd826a216/include/tao/json/basic_value.hpp#L804-L823

**Comment by d-frey** [comment]
Also, seeing that you are using `parse_file`: This should now be `from_file`. Are you using the latest version of our code?

**Comment by gazzatav** [comment]
These are the versions in conan-center:

taocpp-json/1.0.0-beta.11
taocpp-operators/1.2.2
taocpp-pegtl/2.8.1
taocpp-pegtl/2.8.2
taocpp-pegtl/2.8.3
taocpp-pegtl/3.0.0
taocpp-pegtl/3.1.0
taocpp-pegtl/3.2.0
taocpp-sequences/2.0.1
taocpp-taopq/20200222
taocpp-tuple/1.0.0

My project is depending on taocpp-json/1.0.0-beta.11.  should I be using one of the others?

**Comment by d-frey** [comment]
It is a constant struggle to keep things up-to-date in all package managers, and frankly, I've given up on that front. The version in Conan-Center is 3+ years old and although we do have a newer pre-release tag, it doesn't show up there. I can only advise people to use the latest master branch for taoJSON until we will release a real version 1.0.0 - which is now up in the air for several years. Not ideal, but we simply don't have enough time. Anyways, please use the "master" branch and forget about the pre-releases.

**Comment by gazzatav** [comment]
That explains a lot!  If I make a new conan recipe myself then, presumably it will need to depend on PEGL too ( and that will need packaging if the ones in conan-center are out of date).  If I manage to get it working properly  (big if), would you be interested in the new recipes?  Is the latest recipe you have committed?

**Comment by d-frey** [comment]
Usually, @uilianries takes care of Conan-related stuff. Of course, we need to provide regular releases for that to work and this is still a problem with taoJSON. There are a *lot* of things on our to-do list and not enough time to work on those.

That said, I would hope that for all package managers, we, i.e. the authors of open-source libraries, do *not* have to actively get involved in updating packages. There are too many package managers and the process should be automated. If manual work is required the system is already broken IMHO.

We also want to improve CMake support as this might make it easier for people to use the latest version from git directly. Hopefully, this can also take care of the dependency towards the PEGTL. But CMake is quite complicated in itself and again takes a lot of time.

**Comment by gazzatav** [comment]
Yes @uilianries a nice person, very helpful - I bet he's busy though! I did have a look at the conan site and noticed that there is only a conandata.yml for taojson and no config.yml with versions (but I'm a total noob at that, it just struck me as odd).  But you are right according to advice from the folks at cppcon etc,  library authors shouldn't concentrate on packaging, just on making their work packageable.  From my own perspective though, if I find a library in conan that does what I need and has documentation that I can understand I'm going to stick with that and not build every minor version of my project against different versions of the libraries I'm using.

On the point you make about improving CMake support, it always seems to be many versions ahead of where I'm at!  If you couple to PEGTL too tightly with CMake you might make them harder to package as separate entities.  I'm no expert but even I find it relatively easy to edit a CMakeLists.txt to create a package for my local conan cache using export_sources etc.  So then it's trivial to tell conan that one library requires another in that context.
